### PR TITLE
Accessibility and Usability Optimisations

### PR DESCRIPTION
**Pull request by telmiger**
As discussed with Josiah … 
https://github.com/tgrosinger/tw5-checklist/issues/29#issuecomment-304064258

The demo is updated here: https://tid.li/tw5/test/checklist.html 

A big step for this wonderful plugin :)

### CODE REVIEW

File: tiddlers/usage.tid [hunk 1]
```diff
@@ -39,7 +39,3 @@ delimited) checklist items will be displayed as a single checklist.
 List items support any inline formatting rules. This means you can create links,
 bold, or anything else found on the
 [[Formatting in WikiText|http://tiddlywiki.com/static/Formatting%2520in%2520WikiText.html]] guide.
-
```
Comment (tgrosinger) [inline tiddlers/usage.tid]: This should be addressed as part of #30. I would like to not remove this right now because it is a good reminder that it needs to be replaced with a link to the actual documentation.

File: js/checkbox-wikirule.js [hunk 2]
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
...
-    var configWidgetFields = $tw.wiki.getTiddler(configWidgetTitle).fields;
-
-    var strikeChecked = configWidgetFields["strike-checked"] || "true";
-    return (strikeChecked === "true");
-}
 
 exports.shouldShowClearAll = function() {
     var configWidgetTitle = "$:/plugins/tgrosinger/tw5-checklist/Configuration";
     var configWidgetFields = $tw.wiki.getTiddler(configWidgetTitle).fields;
-
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: minimum diff please

File: js/checkbox-wikirule.js [hunk 3]
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
...
     var showClearAll = configWidgetFields["show-clearall"] || "true";
     return (showClearAll === "true");
 }
 
+
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: minimum diff please

File: js/checkbox-wikirule.js [hunk 4]
```diff
@@ -21,30 +21,23 @@ exports.init = function(parser) {
 };
 
 /*
-Retrieve the configuration state indicating if items should be striken
+Retrieve the configuration state of the clear all button
 */
-exports.shouldStrikeChecked = function() {
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js:26]: Why was this removed?

File: js/checkbox-wikirule.js [hunk 5]
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
...
             attributes: {
                 type: {type: "string", value: "checkbox"},
-                pos: {type: "string", value: match.index}
+                pos: {type: "string", value: match.index},
+                id: {type: "string", value: match.index}
             }
         };
         if (match[1] === "x" || match[1] === "X") {
             checkbox.attributes.checked = {type: "boolean", value: true};
-            if (this.shouldStrikeChecked()) {
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js:102]: Why was this removed?

File: js/checkbox-wikirule.js [hunk 6]
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
...
-                itembody.attributes.class.value += ", checkedchecklistitem-body";
-            }
         }
 
-        var removeicon = {
+        var itembody = {
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: Can you move this back up above checkbox? Minimum diff please

File: js/checkbox-wikirule.js [hunk 7]
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
                 this.parser.source.substring(startPos + 4, this.parser.pos),
                 {parseAsInline: true});
 
-        // Put the listitem body in a span for easy reference
-        var itembody = {
-            type: "element",
-            tag: "span",
-            attributes: {
-                class: {type: "string", value: "checklistitem-body"}
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: if checklistitem-body is no longer used it should be removed as part of this commit. Each commit should be a functional intermediate state.

File: js/checkbox-wikirule.js [hunk 8]
```diff
@@ -79,40 +72,54 @@ exports.parse = function() {
...
                 type: {type: "string", value: "checkbox"},
-                pos: {type: "string", value: match.index}
+                pos: {type: "string", value: match.index},
+                id: {type: "string", value: match.index}
             }
         };
         if (match[1] === "x" || match[1] === "X") {
             checkbox.attributes.checked = {type: "boolean", value: true};
-            if (this.shouldStrikeChecked()) {
-                itembody.attributes.class.value += ", checkedchecklistitem-body";
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js:103]: Same as previous comment about checklistitem-body, but for checkedchecklistitem-body.

File: tiddlers/stylesheet.tid [hunk 9]
```diff
@@ -1,6 +1,5 @@
 title: $:/plugins/tgrosinger/tw5-checklist/stylesheet
 tags: $:/tags/Stylesheet
-type: text/css
```
Comment (tgrosinger) [inline tiddlers/stylesheet.tid:3]: Why was this removed?

File: tiddlers/stylesheet.tid [hunk 10]
```diff
@@ -13,24 +12,33 @@ ul.checklist > li {
...
+    left: 1.75em;
+}
+
+input:checked + label.checklistitem-body {
+    color: <<colour muted-foreground>>;
 }
 
-span.checkedchecklistitem-body {
-    color: #ddd;
+<$list filter="[[$:/plugins/tgrosinger/tw5-checklist/Configuration]field:strike-checked[true]]">
```
Comment (tgrosinger) [inline tiddlers/stylesheet.tid]: This looks really cool, can you put a comment on the line above explaining what is happening?

**Review by tgrosinger — COMMENTED** [review]
I haven't quite finished the review yet, but here is what I have so far. Looks really good @telmiger, thank you for putting so much work into this!

File: js/checkbox-wikirule.js (hunk 4 above)
Comment (telmiger) [reply js/checkbox-wikirule.js:26]: We can check this directly from the CSS now, see below. As stirkethrough is a styling issue, I thought this would be the right place.

File: js/checkbox-wikirule.js (hunk 5 above)
Comment (telmiger) [reply js/checkbox-wikirule.js:102]: The stylesheet took over this check. See below.

File: tiddlers/stylesheet.tid (hunk 9 above)
Comment (telmiger) [reply tiddlers/stylesheet.tid:3]: This type does not allow wikitext interpretation, which is used for a) transclusion of color values and b) checking the strikethrough-setting. (No type ist the same as text/vnd.tiddlywiki I believe.)

File: js/checkbox-wikirule.js [hunk 11]
```diff
@@ -79,50 +113,80 @@ exports.parse = function() {
...
+            tag: "label",
+            attributes: {
+                class: {type: "string", value: "checklistitem-body"},
+                for: {type: "string", value: match.index}
+            },
+            children: parseResults.tree
+        };
+
+        // Make a button to delete the item
+        // TODO: use svg bin icon, pattern as follows?
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: Can you please pull this TODO out of the code and open a new issue with this content?

File: js/checkbox-wikirule.js [hunk 12]
```diff
@@ -79,50 +113,80 @@ exports.parse = function() {
...
+    <use xlink:href="#bin-icon"></use>
+  </svg>
+</button>
+*/
+        var removelabel = {
+            type: "element",
+            tag: "span",
+            attributes: {
+                class: {type: "string", value: "checklist-vh"},
+// the next line does nothing, it works here but nowhere else :–(
```
Comment (tgrosinger) [inline js/checkbox-wikirule.js]: What do you mean by this? Why leave the line if it doesn't work?

File: js/checklist.js [hunk 13]
```diff
@@ -56,6 +56,7 @@ CheckListWidget.prototype.execute = function() {
     this.makeChildWidgets();
     this.renderChildren(domNode);
 
+    //  If this is the clear all button …
```
Comment (tgrosinger) [inline js/checklist.js]: Move this down one line, into the each function

File: js/checklist.js [hunk 14]
```diff
@@ -111,13 +113,15 @@ CheckListWidget.prototype.handleClearChecksEvent = function(event) {
 
 // When the user starts typing, change the pencil icon into a checkbox
 CheckListWidget.prototype.handleNewItemTypingEvent = function(event) {
```
Comment (tgrosinger) [inline js/checklist.js]: Why don't we just remove this whole block then instead of leaving in commented-out code

File: tiddlers/stylesheet.tid [hunk 15]
```diff
@@ -1,64 +1,135 @@
...
+
+/* buttons */
+
+button.checklist-add {
+    position: relative;
+    left: 0.75em;
+    width: 18px;
+    height: 18px;
+    border-radius: 5px;
+background-image: url('data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2211%22%20height%3D%2211%22%3E%0A%3Crect%20width%3D%221%22%20height%3D%227%22%20x%3D%225%22%20y%3D%222%22%20fill%3D%22%239B9B9B%22%2F%3E%0A%3Crect%20width%3D%227%22%20height%3D%221%22%20x%3D%222%22%20y%3D%225%22%20fill%3D%22%239B9B9B%22%2F%3E%0A%3C%2Fsvg%3E');
```
Comment (tgrosinger) [inline tiddlers/stylesheet.tid:110]: Is this just a small "+" icon? Can we just use some sort of unicode character instead?

File: tiddlers/stylesheet.tid [hunk 16]
```diff
@@ -1,64 +1,135 @@
...
+}
+
+input:checked + label.checklistitem-body {
+    color: <<colour muted-foreground>>;
 }
 
-div.checklist-removeitem-icon {
-    bottom: 0;
-    color: #777;
+<$list filter="[[$:/plugins/tgrosinger/tw5-checklist/Configuration]field:strike-checked[true]]">
```
Comment (tgrosinger) [inline tiddlers/stylesheet.tid]: Does this just only include the surrounded body if the field is true?

**Review by tgrosinger — CHANGES_REQUESTED** [review]

**Comment by tgrosinger** [comment]
@telmiger Can you put up a an updated version of your demo when you get a chance? Thank you again!

File: js/checkbox-wikirule.js (hunk 11 above)
Comment (telmiger) [reply js/checkbox-wikirule.js]: If we can proceed with my newest version, then this is already obsolete/done.

File: js/checkbox-wikirule.js (hunk 12 above)
Comment (telmiger) [reply js/checkbox-wikirule.js]: I thought this could be a clue for a more elegant solution using (translated) standard texts … but if you don’t know how to make this work either, then I will remove it.

File: js/checklist.js (hunk 13 above)
Comment (telmiger) [reply js/checklist.js]: Got it. I have replaced it with /* add event listeners */ in the upcoming version – the following if-statement speaks for itself.

File: js/checklist.js (hunk 14 above)
Comment (telmiger) [reply js/checklist.js]: Two reasons: 1) I wanted to let you know. 2) This function is called from somewhere and I did not have the time to clean up properly … will do so.

File: tiddlers/stylesheet.tid (hunk 16 above)
Comment (telmiger) [reply tiddlers/stylesheet.tid]: Exactly, you can check the preview/rendered tiddler and you should see the result.

File: tiddlers/stylesheet.tid (hunk 15 above)
Comment (telmiger) [reply tiddlers/stylesheet.tid:110]: We can, if you know a unicode character that looks like a + but is not announced by screenreaders. The recommended way is to use a background image and a button-label. With TW 5.1.14 we have a [more elegant way to create datauri](http://tiddlywiki.com/#datauri%20Macro) – would it be o.k. if the new version is TW 5.1.14+ only?

File: js/checkbox-wikirule.js (hunk 12 above)
Comment (tgrosinger) [reply js/checkbox-wikirule.js]: Ah, no I am not familiar with textReference

File: js/checkbox-wikirule.js (hunk 7 above)
Comment (telmiger) [reply js/checkbox-wikirule.js]: This was still in use, I had replaced the element with a label instead of a span. But you are right, we can remove it and target the label element without a class.

**Comment by telmiger** [comment]
The new version, in which I have tried to consider all your inputs, is online: https://tid.li/tw5/test/checklist.html – I hope you like it. 
If I find out how to update a PR, I will be happy to do so.

**Comment by tgrosinger** [comment]
@telmiger the new demo looks awesome! Once you push the changes to the pull request I will take a final pass over it and then we can get it merged.

Please notice that I made a small change in the Usage.tid which has caused a merge conflict in your review.

**Comment by tgrosinger** [comment]
@telmiger  I just checked out your branch to build the state that has been uploaded, but it is definitely missing changes that are reflected in the demo you uploaded. You should be able to just use a `git push` to upload a new version of your changes.

**Comment by telmiger** [comment]
My master branch is updated. Should I send a new PR or can you proceed with this one here?

**Comment by tgrosinger** [comment]
No need for a new PR, I can see the commits and will take a look at them as soon as I get a chance. Thank you!

File: tiddlers/stylesheet.tid (hunk 15 above)
Comment (tgrosinger) [reply tiddlers/stylesheet.tid:110]: Let's wait to use that until 5.1.14 has been out a bit longer. Want to keep compatibility as flexible as possible

**Review by tgrosinger — APPROVED** [review]
